### ISSUE TITLE
Minimize usages of deasync library in pulumi.

### ISSUE THREAD

**Issue by CyrusNajmabadi** [pulumi/pulumi#3299]
We're seeing a lot of crashes with Mac customers with the following assert: https://github.com/pulumi/pulumi/issues/3260

This is due to https://github.com/abbr/deasync/issues/55, a scary issue which seems very unlikely to be resolved in any sort of reasonable timeframe.

Because of this, and because we do not want to break people who have code sitting on top of our synchronous libs, we need a way to mitigate the impact of this bug as much as we can.

It's worth also noting that the issue we're running into now seems to be OSX only.  As such, we can consider a future were we keep our existing codepaths running exactly like they do today on Windows/Linux and we only follow a new strategy for OSX.  Or we can decide we would like to be as consistent as possible and follow this all the time.  My preference in the latter.

Here are the potential options we can take

### Option 1.  Introduce synchronous invoke path.  With synchronous registration of Provider resources.

This not cheap to implement, but it should ideally reduce the impact of this bug by minimizing the number of cases where the `deasync` codepath gets executed.

1. We resurrect @pgavlin's approach to providing synchronous grpc (sRPC) calls using file-io: https://github.com/pulumi/pulumi/commit/1138b9e4df9def7ea3de92b73dbe3cc0bac35a7d.  This approach does not use any sort of low level hacks (like how `deasync` directly co-opts the nodejs event loop).  But it does require that the file-io code does work properly on all platforms we run on.  We are effectively only doing reads/writes to common shared files between the lang-process and the engine-process.  But we all know that IO can be fraught with issues (esp. on platforms like Windows).  
2. We update Providers to synchronously register (using sRPC) themselves when possible.  This would be possible when all inputs to the provider are simple POJOs.  i.e. no promises or outputs.
    1. Note: we will likely need to update our codegen for providers to support this.  Right now providers *eagerly* become async through code like `inputs["allowedAccountIds"] = pulumi.output(args ? args.allowedAccountIds : undefined).apply(JSON.stringify);`.  We'll likely need to change codegen there to try to avoid going async at all in the provider constructor.
3. We augment Providers to add a bit to them stating if they're registered.  i.e. `isRegistered: bool`.  This is set during synchronous registration, and also set once async-registration is complete if the provider goes the async registration route.
4. Today, all inputs of an `invoke` are required to be POJOs.  However, an `invoke` may be passed in a Provider.   We update `invokes` to synchronously execute (using sRPC) when:
    1. there is no provider, or
    2. the provider is registered. 
    3. in the case where the provider is not registered.  We can either throw (a breaking change), or we can continue calling 'deasync'.  If we make it throw, it will be up to the caller to either pass in `async: true` to the invoke, or do an `await provider.urn`.

This effectively makes it so that we can still hopefully be synchronous *in a safe manner* (with sRPC) for the majority of users in the majority of cases.  The unsafe `deasync` codepath would either be removed (breaking things for a small subset of people), or would remain (possibly with a warning) for what is hopefully a minority of calls/cases.  This would hopefully then happen so much less that the problem we're seeing would be vastly diminished.

### Option 2.  Introduce synchronous invoke path.  Providers must be asynchronously awaited to use htem in invokes.

This is similar to option 1, except we remove all the bits about Providers and their synchronous registration.  Instead, if an Invoke needs to refer to a provider, it must first be the case that someone did something like `await provider.isRegistered`, ensuring that the invoke only happens when it refers to a provider that the engine already knows about.

This is much less work (since we dont' need a synchronous Provider registration path), but pushes more pain onto users to have to `asyncify` their code in order to use invokes and providers.   This approach will work with components that synchronously call `invoke` since the awaiting can happen outside of the component.

### Option 3.  Provide a switch to completely disable synchronous invokes.  

Just as it sounds, we would provide a pulumi config option to no longer automatically perform synchronous invokes.  Instead, all invokes would go back to being async.  At the same time, we would move our existing components off of using the sync invokes and back to the async invokes.

We should likely also catch this specific error being reported and provide a link to a document that helps explain what's going on and what can be done about it.

**Comment by CyrusNajmabadi** [comment]
Tagging @lukehoban @pgavlin .  The above is non-insignificant amount of work (esp. around testing).  Conceptually I think it is sensible.  But I wanted to get everyone on board on this approach before starting anything.

Note: my total preference would be:

1. We do things consistently across all platforms.  So this means getting a file-IO based RPC mechanism working for windows/linux/osx.  I would prefer to void platform-specific code/designs.
2. we update things to be able to avoid `deasync` as much as possible, using the new safe file-io RPC system to be synchronous.
3. in the cases where we can't avoid `deasync` we still call through, but we **emit a warning** with a link to a document explaining what the issue is.

Specifically for '3', i think we can basically warn quite effectively.  Basically, this only happens when a provider is being passed along to an invoke and the Provider is not registered.  In this case, we can advise that users either update their code like so:

```ts
const result = await someDataSource(..., { provider: provider, async: true });

// or
await provider.urn;
const result = someDataSource(..., { provider: provider });
```

It will be painful if this pushes async up to a top-level point that isn't pleasant for their code.  However, this may be less of an issue in the future with work like https://github.com/denoland/deno/pull/3024 and https://chromium.googlesource.com/v8/v8.git/+/0ceee9ad28c21bc4971fb237cf87eb742fc787b8%5E%21/

Basically, both TypeScript and v8 are adding support for top-level awaits.  So, at some point in the future, this may become much more pleasant for users who encounter this warning.

**Comment by pgavlin** [comment]
> Note: we will likely need to update our codegen for providers to support this. Right now providers eagerly become async through code like inputs["allowedAccountIds"] = pulumi.output(args ? args.allowedAccountIds : undefined).apply(JSON.stringify);. We'll likely need to change codegen there to try to avoid going async at all in the provider constructor.

Is there some way to determine whether or not a promise is resolved in JS? Seems like that's all we'd need, right?

**Comment by pgavlin** [comment]
> The unsafe deasync codepath would either be removed (breaking things for a small subset of people), or would remain (possibly with a warning) for what is hopefully a minority of calls/cases

What worries me about this is that it feels relatively common to create k8s providers in Pulumi using the outputs of another resource as inputs. In practice this may not be much of an issue, as k8s providers aren't typically used with invokes (indeed, I think the provider only has one invoke, if any). I wonder if there's any way for us to get data on this.

**Comment by pgavlin** [comment]
I do think that this plan seems reasonable overall. From an SDK usability perspective, the thing that I find most worrisome is the difficulty of accommodating necessarily-async calls inside `ComponentResource` constructors--what approach would you take there?

**Comment by CyrusNajmabadi** [comment]
> Is there some way to determine whether or not a promise is resolved in JS? Seems like that's all we'd need, right?

There is not :(

**Comment by CyrusNajmabadi** [comment]
> What worries me about this is that it feels relatively common to create k8s providers in Pulumi using the outputs of another resource as inputs. In practice this may not be much of an issue, as k8s providers aren't typically used with invokes (indeed, I think the provider only has one invoke, if any). 

Right.  So as a *mitigation* strategy, I have more confidence about what is outlined here.  For the common case of an *async* Provider (i.e. k8s) there is less chance of running into an issue since invokes will not be involved.  For cases where we have invokes, we likely have synchronous Providers.  

So, overall, the likelihood that we hit `deasync` is greatly diminished.

**Comment by CyrusNajmabadi** [comment]
> the thing that I find most worrisome is the difficulty of accommodating necessarily-async calls inside ComponentResource constructors--what approach would you take there?

In the short term, the approach we're taking here greatly mitigates this issue.  Components can still just call the synchronous invokes and callers will either:

1. not being using providers. in which case things will work since we will just use the new sRPC invoke.
2. be using synchronously registered providers.  in which case, things will work since we can use the new sRPC invoke.
3. be using async registered provider.  This will likely be ok for some subset of users, and not ok for others.  For those that still have problems here, a (painful) workaround will be for them to call `await provider.isRegistered` after they make their provider.  This will require them to async'ify their code, which is not pleasant.

--

In the medium term we could consider a more impactful (i.e. breaking) change that basically mandates what is discussed in '3' above.  Specifically, users would have to `await` "something" about the Provider before making any calls.  One approach do to this would be:

1. remove `Provider` from all our options bags.  Instead, require a `ProviderRef` (this is already breaking).
2. Users can get such a ref like so `await provider.getRef()`, forcing their hand to do the async work before calling into anything.
3. At this point, 100% of invokes could be synchronous using sRPC.  deasync can be removed entirely.
4. We would be able to remove synchronous provider registration codepaths.

**Comment by CyrusNajmabadi** [comment]
Talked to @lukehoban today and came up with an `option 3` in the main post.  @pgavlin the gist of this is that we will:

1. Have a global opt-out switch for synchronous invokes.  Users can set this and all invokes will be async.
2. We will update our components to always use async invokes.  In the one place we can't (a couple of VPC static factory methods), we will provide an async entrypoint as well.  
3. We can try to trap this exception that node is throwing and provide a better message/link to documentation explaining what can be done here.

This has the following benefits:

1. users who are not having problems continue to have no problems. 
2. users who have this problem have a solution they can apply.  Thee solution is a break, and will be painful, but can be isolated to that set of users.

**Comment by CyrusNajmabadi** [comment]
Now that we've made decisions, i'm closing this out so we can instead track things in: https://github.com/pulumi/pulumi/issues/3318
